include/pybind11/chrono.h

```diff
@@ -0,0 +1,177 @@
+                                          , localtime.tm_mon + 1
+                                          , localtime.tm_mday
+                                          , localtime.tm_hour
+                                          , localtime.tm_min
+                                          , localtime.tm_sec
+                                          , (duration_cast<us_t>(src.time_since_epoch() % seconds(1))).count());
+    }
+    PYBIND11_TYPE_CASTER(type, _("datetime.datetime"));
+};
+
+// Other clocks that are not the system clock are not measured as datetime.datetime objects
+// since they are not measured on calendar time. So instead we just make them timedeltas
+// Or if they have passed us a time as a float we convert that
+template <typename Clock, typename Duration> class type_caster<std::chrono::time_point<Clock, Duration>> {
+public:
+    typedef std::chrono::time_point<Clock, Duration> type;
+    typedef std::chrono::duration<std::chrono::hours::rep, std::ratio<86400>> days;
+
+    bool load(handle src, bool) {
+        using namespace std::chrono;
+        if (!PyDateTimeAPI) { PyDateTime_IMPORT; }
+
+        // If they have passed us a datetime.delta object
+        if (PyDelta_Check(src.ptr())) {
+            // The accessor macros for timedelta exist in some versions of python but not others (e.g. Mac OSX default python)
+            // Therefore we are just doing what the macros do explicitly
+            const PyDateTime_Delta* delta = reinterpret_cast<PyDateTime_Delta*>(src.ptr());
+            value = time_point<Clock, Duration>(
+                  days(delta->days)
+                + seconds(delta->seconds)
+                + microseconds(delta->microseconds));
+            return true;
+        }
+        // If they have passed us a float we can assume it is seconds and convert
+        else if (PyFloat_Check(src.ptr())) {
+            double val = PyFloat_AsDouble(src.ptr());
+            value = time_point<Clock, Duration>(Duration(std::lround((val / Clock::period::num) * Clock::period::den)));
+            return true;
+        }
+        else return false;
+    }
+
+    static handle cast(const std::chrono::time_point<Clock, Duration> &src, return_value_policy /* policy */, handle /* parent */) {
+        using namespace std::chrono;
+
+        // Lazy initialise the PyDateTime import
+        if (!PyDateTimeAPI) { PyDateTime_IMPORT; }
+
+        // Declare these special duration types so the conversions happen with the correct primitive types (int)
+        using dd_t = duration<int, std::ratio<86400>>;
+        using ss_t = duration<int, std::ratio<1>>;
+        using us_t = duration<int, std::micro>;
+
+        Duration d = src.time_since_epoch();
+
+        return PyDelta_FromDSU(
+              duration_cast<dd_t>(d).count()
+            , duration_cast<ss_t>(d % days(1)).count()
+            , duration_cast<us_t>(d % seconds(1)).count());
+    }
+    PYBIND11_TYPE_CASTER(type, _("datetime.timedelta"));
+};
```

Comment:
Unless I'm missing something, the `load` and `cast` of this type_caster template and the first one (for `std::chrono::duration`) is almost entirely duplicated code: `load` and `cast` across the two types only seem to differ meaningfully on one line (since one is a time point containing a duration, the other is a duration directly).  It would be very nice to put these into a pair of templated helper function to avoid this duplication.  Perhaps something like the following (untested, and you can probably improve upon it):

```C++
template <typename Duration, typename Period> Duration duration_loader(Duration &value, handle src) {
    // ...
}
template <typename Duration> duration_caster(const Duration &src) {
    // ...
}

    // In the std::chrono::duration type caster:
    bool load(handle src, bool) { return duration_loader<type, type::period>(value, src); }
    static handle cast(const std::chrono::duration<Rep, Period> &src, return_value_policy /* policy */, handle /* parent */) {
        return duration_caster(src);
    }

    // and in the generic clock type caster:
    bool load(handle src, bool) {
        Duration dur;
        if !(duration_loader<type, type::period>(dur, src)) return false;
        value = time_point<Clock, Duration>(dur);
        return true;
    }
    static handle cast(const std::chrono::time_point<Clock, Duration> &src, return_value_policy /* policy */, handle /* parent */) {
        return duration_caster(src.time_since_epoch());
    }
```